src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,763 @@
+  _sampler->start_thread();
+  _sampler->enroll();
+}
+
+void JfrCPUTimeThreadSampling::update_run_state(double rate, bool autoadapt) {
+  if (rate != 0) {
+    if (_sampler == nullptr) {
+      create_sampler(rate, autoadapt);
+    } else {
+      _sampler->set_rate(rate, autoadapt);
+      _sampler->enroll();
+    }
+    return;
+  }
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate /* 0 */, autoadapt);
+    _sampler->disenroll();
+  }
+}
+
+void JfrCPUTimeThreadSampling::set_rate(double rate, bool autoadapt) {
+  assert(rate >= 0, "invariant");
+  if (_instance == nullptr) {
+    return;
+  }
+  instance().set_rate_value(rate, autoadapt);
+}
+
+void JfrCPUTimeThreadSampling::set_rate_value(double rate, bool autoadapt) {
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate, autoadapt);
+  }
+  update_run_state(rate, autoadapt);
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_create(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_create(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_terminate(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_terminate(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::trigger_async_processing_of_cpu_time_jfr_requests() {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->trigger_async_processing_of_cpu_time_jfr_requests();
+  }
+}
+
+void handle_timer_signal(int signo, siginfo_t* info, void* context) {
+  assert(_instance != nullptr, "invariant");
+  _instance->handle_timer_signal(info, context);
+}
+
+
+void JfrCPUTimeThreadSampling::handle_timer_signal(siginfo_t* info, void* context) {
+  assert(_sampler != nullptr, "invariant");
+  if (Atomic::load(&_sampler->_stop_signals)) {
+    return;
+  }
+  Atomic::inc(&_sampler->_active_signal_handlers, memory_order_acq_rel);
+  _sampler->handle_timer_signal(info, context);
+  Atomic::dec(&_sampler->_active_signal_handlers, memory_order_acq_rel);
+}
+
+void JfrCPUTimeThreadSampler::sample_thread(JfrSampleRequest& request, void* ucontext, JavaThread* jt, JfrThreadLocal* tl) {
+  JfrSampleRequestBuilder::build_cpu_time_sample_request(request, ucontext, jt, jt->jfr_thread_local());
+}
+
+volatile size_t count__ = 0;
+
+static bool check_state(JavaThread* thread) {
+  switch (thread->thread_state()) {
+    case _thread_in_Java:
+    case _thread_in_native:
+      return true;
+    default:
+      return false;
+  }
+}
+
+void JfrCPUTimeThreadSampler::handle_timer_signal(siginfo_t* info, void* context) {
+  JavaThread* jt = get_java_thread_if_valid();
+  if (jt == nullptr) {
+    return;
+  }
+  JfrThreadLocal* tl = jt->jfr_thread_local();
+  JfrCPUTimeTraceQueue& queue = tl->cpu_time_jfr_queue();
+  if (!check_state(jt) ||
+      jt->is_JfrRecorder_thread()) {
+      queue.increment_lost_samples();
```

Comment:
is_JfrRecorder_thread() will not appear here since it's excluded and would have returned nullptr from get_java_thread_if_valid().